Savestate speedups and audio fixes

Dwedit opened the pull request:
* Saves and loads the audio buffer state for clean audio at 0 runahead
* Clean audio for Sega CD PCM when loading state
* Clean audio for CD Audio when loading state by saving the sample position.  Supported for CD images, Vorbis tracks, and CHD images.
* Loading state no longer invalidates tile cache, so emulator doesn't have to recreate it after loading state, big speedup
* Hard disable audio support cleaned up and re-added.  Now it uses some dummy callbacks to replace the FM synthesis.
* Fixed Savestates saved or loaded during TMSS bootrom
* MSVC project updated, featuring full libchdr support
* Removed the ".user" file from the MSVC project, added to git ignore list

ekeeke commented:
Once again, this looks very hackish and I would have rather prefered to review each core changes myself before this got automatically accepted by libretro devs with some explanations of the underlying needs/issues so I could eventually propose an alternate and better solution to whatever your problem is, so it could eventually benefit to any other ports, not just libretro.

As already said, there are simpler and cleaner ways to get fast state loading (and no reason to make it conditional either) and likely cleaner ways to get clean audio on state load (although I never had any issues with it before so I'm curious why exactly it is a problem now ?). As for audio disable, again I am not sure to understand why exactly this is needed by Retroarch but what I know is that disabling FM synthesis will cause some issues in games that polls the FM timers flags in FM status to trigger some action and will certainly change the state of said games. Same goes for any sound chip that is readable from software, like Mega CD PCM chip for exampl. Sure you are saving some processing speed but you are also making the emulation inaccurate, so why not go further and also makes VDP emulation faster by making it less accurate since frame is skipped anyway... or just use a faster and less accurate emulator? 

In this state, I'm afraid I will never accept these changes upstream so, unless a cleaner solution is found instead of somehow patching main core modules to get some speedup or fix some edge use cases,  the libretro fork will become an unofficial fork by now, which is kinda unfortunate (and will make any real emulation fixes made to upstream harder for you to backport).

twinaphex commented:
This has nothing to do with accuracy or emulation. This has to do with speed fixes for making runahead latency reduction feasible with this core without the performance being disastrous.

twinaphex commented:
I can meet you halfway and we can have a non-shallow fork Genesis Plus GX libretro core where we make edits like this and a shallow fork one that we leave unaltered, like this was before. If you are satisfied with that, we can go ahead and do that.

Beyond that, I don't know what we can do. We care about certain features that apparently you don't care about and it would take months or even years for you to finally cater to them, we cannot wait that long. So either you are receptive to these things in the mainline version or we have to figure out an alternative core for these things.

I don't want this to get into a headbutting or a clash. At the same time, I also don't want to slam the door shut on perfectly fine great new features just because one side has trouble understanding why we need it to begin with. That kind of filibustering is not how I run my project.

ekeeke commented:
No it wasn't explained, just as nowhere it was explained why exactly audio disable is needed. I can not provide any solution if I am not made aware of the problem/need in the first place and this stays hidden in libretro repository (I only noticed it while looking recently for getting both repo back in sync before comitting some changes).

The only thing I am criticizing is the way those changes were implemented (which is very hacky in my opinion and can lead to beneath issues) and how core changes were accepted without any review from people who maintains it. I indeed believe I am more competent than you to know what could possibly break emulation or introduce unwanted issues, no matter how much you think speed hacks are good for your users, just as I think I know this emulator better than you both do, and could propose simpler and cleaner solutions to your needs.

hizzlekizzle commented:
I don't think this needs to be an argument. If Eke comes up with a cleaner, better solution, great! We can always revert this patch / rebase against upstream.

ekeeke commented:
I still need Dwedit to explain me in detail what are the issues/needs he was attempting to fix.

twinaphex commented:
We can rename this non-shallow fork to genesis_plus_gx_experimental for now, in a similar way to how there is a 'Canary' branch of Citra.

twinaphex commented:
> So why don't you try to reach me out first before saying it will take too long or I won't accept core changes ? Like I said, anything that could benefit to other ports is interesting to me.

I did reach out to you I believe. I think I even reached out to you about Atgames some time ago by sending you an email. I just never got a response. I did get a response from Charles though.

Anyway, I mean no harm here, and Dwedit means no harm. We can consider this current fork an 'experimental' fork until you and Dwedit figure out a way to implement this in mainline, then we rebase. Sounds agreeable?

ekeeke commented:
Sounds good to me, as this is similar to how CHD, overclock and sprite limit removal supports were handled.

fr500 commented:
please don't make another fork, just make a branch...
anyway I can try to explain.

Audio and Video disable are used for the runahead feature, basically he doesn't need the video / audio portions when running a few frames ahead so he's running without that, saving a state and rolling back when required.

It's basically a hack to lower the games built-in input lag (reducing the overall perceived input lag).
These changes allow faster savestates with runahead improving performance.

Dwedit commented:
Runahead has very different requirements for save state accuracy than normal emulation.  When you are saving and loading state 60 times per second, every inaccuracy adds up.

I can go through all the changes I made one-by-one.  If you have any questions, I'll be happy to answer them.

First off, there was saving and loading the sound buffer.  This part is very hacky as currently implemented, and I'd be happy to change the way it is done.  The way I implemented saving and loading the sound buffer may be questionable.  Rather than adding a public Serialize/Deserialize function, I maintained the existing code structure of using opaque structs, and added in another opaque struct for holding the sound buffer state.  I understand that may be silly, and will gladly replace that with a proper serialize/deserialize function.
During normal emulation, and normal savestates, you don't really care about the sound buffer's state. 
 But if you're saving state and loading state unconditionally 60 times per second, it does matter.  You can confirm accuracy by logging the audio output.  If you save and load audio buffer state, your wave output will match identically.

As for CD audio state, it simply wasn't accurately storing the sample position in the savestate.  It was accurate to a sector number only, and even that sector number differed from what was actually playing.  If you loaded 60 times per second, there was a lot of nasty noise without recording the sample offset.
I wanted to avoid breaking state compatibility, so I packed the Sample Offset into the oversized field previously used for the current track number.

Then there's no longer invalidating the tile cache on load state.  This is simply a big performance boost with no downsides.

Then there's fixing savestates made during the TMSS screen, which is an issue whether there's runahead or not.  For non-runahead play, you can simply not do savestates for the first few seconds, but runahead is always saving and loading state, so it's important to have those things fixed.

As for hard disable audio, I thought that making FM dummy functions would be the least invasive way to add that feature into the emulator.  There is a real 10% performance boost to skipping audio emulation during frames where it will be discarded.
For the other audio generators (PSG, CD audio, etc..), there weren't any less-invasive ways to add hard disable audio than adding a condition and early return.

Is FM synthesis status even readable?  From what I saw, reading back the status is not even available on all Genesis FM chips, and I don't know of any existing programs that do read back status.  I also see no way for PSG status to be read back either.  Sega CD PCM status can be read back however, so I wrote a silent version of the code that increments the frequency counters, loops them, and such, and outputs no audio.

Anyway, there are issues with how the savestate format works right now.  Since there are no Chunks (four-byte name then chunk size), you can't add any bytes anywhere without invalidating all savestates.  Additionally, savestates made on a big-endian system can not be loaded on a little-endian system.  Also, aligning 32-bit words to 32-bit boundaries would improve performance slightly, and add compatibility for old ARM processors which can't read unaligned words.

ekeeke commented:
> Runahead has very different requirements for save state accuracy than normal emulation. When you are saving and loading state 60 times per second, every inaccuracy adds up.

Could you briefly explain what is done exactly in regard to frame emulation, skipped and rendered frames: when frame state is saved ? when it is loaded ? I have a vague understanding how this run-ahead thing is working but I am not exactly sure about the events chain.

>First off, there was saving and loading the sound buffer. This part is very hacky as currently implemented, and I'd be happy to change the way it is done. The way I implemented saving and loading the sound buffer may be questionable. Rather than adding a public Serialize/Deserialize function, I maintained the existing code structure of using opaque structs, and added in another opaque struct for holding the sound buffer state. I understand that may be silly, and will gladly replace that with a proper serialize/deserialize function.
During normal emulation, and normal savestates, you don't really care about the sound buffer's state.
But if you're saving state and loading state unconditionally 60 times per second, it does matter. You can confirm accuracy by logging the audio output. If you save and load audio buffer state, your wave output will match identically.

The thing is, blip buffers are filled during the frame then emptied at the end of the frame (twhen they are read into the frontend sound buffer) and normally, there shouldn't be any samples (or more exactly sample deltas as this is how blip buffers work for audio resampling) left in those blip buffers (PSG and FM are using the same blip buffer since they share the same master clock / timing base and all the available samples at the end of the frame are resampled, while audio CD and PCM buffers are normally filled up to the needed number of samples, based on the first blip buffer returned number of samples). Can you give some examples of issues you encountered without blip buffer saving?

As for the last buffer values, I guess it depends what you are exactly doing with frame states and what frames you are skipping/rendering but they are only used to know the first sample delta value (difference between the last sample output from previous frame and the first sample of upcoming frame) so it would seem more logical to me to use the last values of the rendered frame just before loading the state. Again, was there any issues fixed specifically fixed by this addition? 

Maybe it would also be worth to retry with system_reset call removed from state load since it also resets blip buffers and might indirectly cause these issues you encountered. I will try anyway to propose a better solution for faster state loads without this overall reset so you can test it. If it still is not enough, we can have a look at this buffer saving thing and clean it a little bit.

>As for CD audio state, it simply wasn't accurately storing the sample position in the savestate. It was accurate to a sector number only, and even that sector number differed from what was actually playing. If you loaded 60 times per second, there was a lot of nasty noise without recording the sample offset.
I wanted to avoid breaking state compatibility, so I packed the Sample Offset into the oversized field previously used for the current track number.

I understand the required change and there is indeed a problem if you want audio file playback to continue exactly where it stopped (for continuous audio playback). But I think I can find a better solution where only the load and save_context functions are modified, and the whole CDD emulation process (even if those changes are minor, it's not very optimal to have functions updating offsets that only serves for state load). As for breaking savestate compatibility, I think this is unavoidable when you are improving savestates.

> Then there's no longer invalidating the tile cache on load state. This is simply a big performance boost with no downsides.

Yes, this one was indeed a good idea but again, I think it could be made better and simpler without creating extra functions for this in the VDP core and by reusing existing macro. It could also be improved by taking into account the fact the tile cache can be invalidated for each lines of a given tile separately, and the fact tile cache is smaller in mode 4 vs mode 5.

>Then there's fixing savestates made during the TMSS screen, which is an issue whether there's runahead or not. For non-runahead play, you can simply not do savestates for the first few seconds, but runahead is always saving and loading state, so it's important to have those things fixed.

This one looks ok for me and it is indeed fixing an unsupported edge case. Problems remain however  with other edge cases that are saving/loading state with Master System (and posdibly Game Gear) boot ROM. I know what the problem is and how to fix though so I should be able to commit some changes regarding this.

>As for hard disable audio, I thought that making FM dummy functions would be the least invasive way to add that feature into the emulator. There is a real 10% performance boost to skipping audio emulation during frames where it will be discarded.
For the other audio generators (PSG, CD audio, etc..), there weren't any less-invasive ways to add hard disable audio than adding a condition and early return.

That is the main thing I don't understand. If you skip audio chip emulation, it means their internal state is not updated and so when you start rendering video/audio again, the sound chips will be late related to CPU state (i.e registers writes occured within CPUs but the chip state did not changed). So how could this possibly work ? Maybe I'm missing something here about what is done with these skipped frames...

Anyway, I think that before thinking disabling audio chip emulation alltogether, we should first test with more optimized state loading, then with disabled blip buffer callbacks (using configurable function pointers for example, which could also serve switching between blip buffer resampling quality). 

Also I mean, there will always be a hard limit for speed and it is highly emulator dependent so you have to put a limit somewhere when ypu optimize things. If people can not use the run ahead feature with Genesis Plus GX on slow platforms then so be it, there are faster / more optimized emulators for them to try, with different accuracy goals. From what I remember, it runs more or less around 200 fps on Wii and so is capable to run more than 3 internal frames per output frame.

>Is FM synthesis status even readable? From what I saw, reading back the status is not even available on all Genesis FM chips, and I don't know of any existing programs that do read back status.
There is a single FM chip in Genesis (with two variants depending on used technology but interface is always the same) and off course it is readable, why would there be a YM_Read function in the emulator otherwise ? Not sure what you have been looking at but most games (if not all) read the FM busy flag from status register before writing the chips and many games read the timer flags in their sound engine to synchronize with the chip (there is no other way as IRQ pin is not connnected).

> I also see no way for PSG status to be read back either.
No, that's why I didn't mentionned it. The chip still runs synchronized with 68k and Z80 CPus though and performs things internally like noise update or tone level shifting based on the programmed frequencies and elapsed time. Again, if you don't update the internal state when update function is called at the end of the frame, the internal state will stay where it was the time the chip was written, which could have happen several frames before. But maybe this is expected to you and taking in account by something else, I am not sure at this point.

> Sega CD PCM status can be read back however, so I wrote a silent version of the code that increments the frequency counters, loops them, and such, and outputs no audio.
So here PCM chip remains updated and synchronized to CPU state (unlike other sound chips) but rather than again adding a specific function in PCM core for something that is merely an edge frontend use-case (also kinda specific to retroarch actually), I would at worst prefer to have a blip buffer function pointer doing nothing to save resampling processing time.

>Anyway, there are issues with how the savestate format works right now. Since there are no Chunks (four-byte name then chunk size), you can't add any bytes anywhere without invalidating all savestates. Additionally, savestates made on a big-endian system can not be loaded on a little-endian system. Also, aligning 32-bit words to 32-bit boundaries would improve performance slightly, and add compatibility for old ARM processors which can't read unaligned words.

I agree with you for the first one and already had various ideas in mind to solve this. The second one (portability) is hardly due to the format but rather to the state load function, while the last one is more an optimization for memcpys on some specific platforms than really a compatibility issue. Anyway, surely the state format is not perfect but it never was meant to be and all those things are more enhancement than real issues preventing states to be correctly reloaded manually during the game, which is honestly the original goal of this feature.

Dwedit commented:
> Could you briefly explain what is done exactly in regard to frame emulation, skipped and rendered frames: when frame state is saved ? when it is loaded ? I have a vague understanding how this run-ahead thing is working but I am not exactly sure about the events chain.

Here's a brief explanation of how I'm doing RunAhead right now:

Basic RunAhead:
* Run frame (audio and video disabled)
* Save State
* Run X - 1 frames (audio and video disabled)
* Run frame (audio and video enabled)
* Load State

Video Disabled is handled simply by passing the discard frame flag to the run function.
Audio Disabled is handled by having the RetroArch audio callback discard the generated audio samples.  We still want to run all the audio generation as normal in this case.

RunAhead with Secondary core without resync:  (Despite being called "secondary core", this is all single-threaded, but this would be a good opportunity to push one core into a different thread.)
* Run frame (audio enabled, video disabled)
* Check if input is dirty or not, if it's dirty, do a resync
* Otherwise: Secondary core: Run frame (video enabled, audio disabled)

RunAhead with Secondary core doing a resync:
* Run frame (audio enabled, video disabled)
* Save state
* Secondary core: Load state
* Secondary core: Run X - 2 frames (audio and video disabled)
* Secondary core: Run frame (video enabled, audio disabled)

Here, the secondary core never needs to output audio.  So we use Hard Audio Disable to get a small performance boost.  Audio is never needed again once Hard Audio Disable has been engaged, and we won't be saving state after using Hard Audio Disable.

One alternative way to do RunAhead makes use of Hard Audio Disable to try to speed things up.  I was considering this for the slower systems, such as Wii, Xbox, PS3, etc...  I'm not currently using this yet.
* Run frame (audio enabled, video disabled)
* Save state
* Run X - 1 frames (Hard audio disable, video disabled)
* Run frame (Hard audio disable, video enabled)
* Load state

So that's the summary of how RunAhead works.

> Blip buffers are filled during the frame then emptied at the end of the frame...there shouldn't be any samples left in those blip buffers when a state is saved.

Blip buffers always have 16 samples remaining inside of them after ending a frame.  There is also the time offset and accumulator.

>It would seem more logical to me to use the last values of the rendered frame just before loading the state.

The 16 samples need to be the correct samples, and not samples from another frame.  The time offset and accumulator also need to be the correct values, and not from another frame.

>Maybe it would also be worth to retry with system_reset call removed from state load since it also resets blip buffers

Skipping a sound buffer reset might save a memset, but it won't provide the correct data needed for clean audio on the next frame.

Only saving and loading the 16 samples, time offset, and accumulator will provide clean audio.

> I understand the required change and there is indeed a problem if you want audio file playback to continue exactly where it stopped (for continuous audio playback). But I think I can find a better solution where only the load and save_context functions are modified, and not the whole CDD emulation process (even if those changes are minor, it's not very optimal to have functions updating offsets that only serves for state load). As for breaking savestate compatibility, I think this is unavoidable when you are improving savestates. It seems like the version was bumped anyway in your pull request.

Before, there was a hidden sample offset.  It was buried inside of the file handle that was reading the cd audio data. 
So I exposed this status as a Sample Offset instead of a File Offset.  That way it does not care about the exact format of the CD image, whether it's compressed, uses FLAC, uses Vorbis, etc.
Already, the sample offset (or file offset from reading the CD track data) was way out of sync with the LBA values that got stored in the savestate.
So I decided that CD audio track position would instead be decided entirely by Sample Offset, rather than the LBA status that did not reflect what the playback code was actually doing.  Using Sample Offset also makes it easier to handle compressed audio.

I bumped the version number up because changing a 32-bit field to an 8-bit and 24-bit field allowed backward compatibility, but an old version would try to read a horribly out of range track number if it loaded such a savestate.

>Yes, [not invalidating tile cache] was indeed a good idea but again, I think it could be made better and simpler without creating extra functions for this in the VDP core and by reusing existing macro. It could also be improved by taking into account the fact the tile cache can be invalidated for each lines of a given tile separately, and the fact tile cache is smaller in mode 4 vs mode 5.

I had encountered some problems when I tried using the macros alongside 32-bit Memcmp blocks, I saw only the top lines of the tiles getting updated, so I changed it to just write FF values instead.  Maybe doing a more simple byte-by-byte comparison and using the macros would make the multiple lines method work.   I'm kinda stuck in a ARM7 mindset where using memory block functions (memset, memcpy, memcmp, etc) are much faster than using byte-by-byte loops, so I was trying to avoid byte-by-byte comparisons.

>There is a single FM chip in Genesis (with two variants depending on used technology but interface is the same) and off course it is readable, why would there be a YM_Read function in the emulator otherwise ? Not sure what you have been looking at but most games (if not all) read the FM busy flag from status register before writing the chip and many games read the timer flags in their sound engine to synchronize with the chip (there is no other way as IRQ pin is not connected).

I had completely missed the busy flag stuff, thanks for the heads up.  Yeah, I think I completely screwed up in assuming the FM chip was not readable.  It just happens to always return not busy, so games are working.  I'd probably need to redo the FM skipping stuff now.

>Anyway, I think that before thinking disabling audio chip emulation alltogether, we should first test with more optimized state loading, then with disabled blip buffer callbacks (using configurable function pointers for example, which could also serve switching between blip buffer resampling quality).

Did the C version of the blip code use callbacks?  I didn't notice them anywhere, and still don't see any callbacks in there.  When I modified QuickNES, I just set it to the "Slient buffer", and that magically worked, since it was c++ based and used callbacks.

================

One other thing, it wouldn't be very hard to add RunAhead support to the native versions of Genesis Plus GX.  Since the RunAhead algorithm itself is so simple, it can be done in like 5 lines of code.

ekeeke commented:
>Here's a brief explanation of how I'm doing RunAhead right now:

Ok so hard audio disable is only used when using a second core instance but I don't really understand what is the benefit of using that second core instance. What is the problem with basic run-ahead ? Also, it is not very clear how the two cores interacts, what do you mean by 'resync' and what is the exact chronology of execution of both core instances ? Could you show what happen exactly on a two frames period (I mean host frames i.e 2x1/60s for example) ?
I am still wondering about the possibility that sound chips state become desynchronized from CPU states when audio is hard disabled.

> One alternative way to do RunAhead makes use of Hard Audio Disable to try to speed things up. I was considering this for the slower systems, such as Wii, Xbox, PS3, etc... I'm not currently using this yet.
> 
>     Run frame (audio enabled, video disabled)
>     Save state
>     Run X - 1 frames (Hard audio disable, video disabled)
>     Run frame (Hard audio disable, video enabled)
>     Load state

The problem here is that audio and video get desynchronized (if the game is triggering a jump sound in sync with the jump animation, it will now be heard X frames after the jump occurred on screen).

> Blip buffers always have 16 samples remaining inside of them after ending a frame. There is also the time offset and accumulator.

There is indeed that `buf_extra thing` (which should be equal to 18 samples though, not 16) but I would have thought it did not held any needed sample for the next frame. I might be wrong though if that makes a difference. Also, looking a bit more into this, it's actually possible for PCM and CD-DA to generate a bit more samples than needed because of resampling ratios rounding, so it seems state should indeed store remaining buffer samples (an easy way would be to add a remain field in blip structure that get updated when `remove_samples` is called. 

> The 16 samples need to be the correct samples, and not samples from another frame. The time offset and accumulator also need to be the correct values, and not from another frame.
I wasn't talking about blip structures but rather last sample values for FM (fm_last[]) and CD-DA (cdd.audio[2]) but they obviously need to be saved as well (it is done for PSG and PCM soundchips already so it should not be any different for these chips).

>Skipping a sound buffer reset might save a memset, but it won't provide the correct data needed for clean audio on the next frame.
>
>Only saving and loading the 16 samples, time offset, and accumulator will provide clean audio.

And FM/CD-DA last sample values also I guess as mentioned above. I will try to improve the *_context functions on that regard (and add specific context load/save functions for audio in system.c that will call blip_buffer context load/save functions). These should be called in state_load / state_save functions in my opinion (better have cleanest state as possible by default), although I am not sure why this was separated in your pull request (blip buffer save/restore is called by libretro). Is there a reason for this ?

> Before, there was a hidden sample offset. It was buried inside of the file handle that was reading the cd audio data.
> So I exposed this status as a Sample Offset instead of a File Offset. That way it does not care about the exact format of the CD image, whether it's compressed, uses FLAC, uses Vorbis, etc.
> Already, the sample offset (or file offset from reading the CD track data) was way out of sync with the LBA values that got stored in the savestate.
> So I decided that CD audio track position would instead be decided entirely by Sample Offset, rather than the LBA status that did not reflect what the playback code was actually doing. Using Sample Offset also makes it easier to handle compressed audio.
> I bumped the version number up because changing a 32-bit field to an 8-bit and 24-bit field allowed backward compatibility, but an old version would try to read a horribly out of range track number if it loaded such a savestate.

Yes, I understand this (and I know LBA is not synced to file offset, it's design choice) but what I meant is that I prefer to find a way to compute that file offset (or sample offset if simpler) only in context functions, and not have other functions increment or reset a global values anytime file is open,read or seeked)

>I had encountered some problems when I tried using the macros alongside 32-bit Memcmp blocks, I saw only the top lines of the tiles getting updated, so I changed it to just write FF values instead. Maybe doing a more simple byte-by-byte comparison and using the macros would make the multiple lines method work. I'm kinda stuck in a ARM7 mindset where using memory block functions (memset, memcpy, memcmp, etc) are much faster than using byte-by-byte loops, so I was trying to avoid byte-by-byte comparisons.

I was more thinking about directly comparing unsigned ints using pointers. This is also were aligned fields would become needed I guess. I will make an attempt in a test branch for you to test it. 

>I had completely missed the busy flag stuff, thanks for the heads up. Yeah, I think I completely screwed up in assuming the FM chip was not readable. It just happens to always return not busy, so games are working. I'd probably need to redo the FM skipping stuff now.

The busy flag is not emulated in most emulators so it's not really a problem. The problem is the FM Timers flags which are needed by CPU to sync with the chip . It's unlikely games use these to sync VDP processing though (but you can never be sure).

>Did the C version of the blip code use callbacks? I didn't notice them anywhere, and still don't see any callbacks in there. When I modified QuickNES, I just set it to the "Slient buffer", and that magically worked, since it was c++ based and used callbacks.

No it doesn't (yet). But that's the cleanest way to do this in my opinion and a first step to attempt before totally disabling sound chip emulation. Have you tried to simply disable settings that have impact on audio processing (config.hq_fm, config.hq_psg and config.filter)  to see if the gain was substantial enough for slow platforms? 
I understand the desire to save the most processing time as possible and going to extreme solution immediately but intermediate solutions should be studied first (and in last resort, a more elegant and global solution could be studied to disable or simplify sound chip emulation rather than patching that global data test in every functions) What is your benchmarks by the way ?
